python-backend/app/auth.py

```diff
@@ -0,0 +1,65 @@
+import re
+
+from cached_property import cached_property
+from flask import g, request
+from uuid import UUID
+from typing import Mapping, Optional, Set
+from .api.utils.include.user_info import User
+from .api.users.minespace.models.minespace_user import MinespaceUser
+from app.api.utils.access_decorators import MINESPACE_PROPONENT
+
+
+class UserSecurity(object):
+    def __init__(self, user_id: int, access: Optional[Set[UUID]] = None):
+        self.access = access or {}
+        self.user_id = user_id
+
+    def __repr__(self):
+        return "<{} mine_ids={}>".format(type(self).__name__, self.mine_ids)
+
+    def is_restricted(self):
+        return get_user_is_proponent()
+
+    @cached_property
+    def mine_ids(self) -> Set[UUID]:
+        if not self.user_id:
+            return None
+
+        return get_mine_access()
+
+    def get_permission(self, mine_id: UUID):
+        return self.access.get(mine_id)
+
+
+def get_mine_access():
+    user = get_current_user()
+    return list(x.mine_guid for x in user.mines)
+
+
+def get_current_user():
+    rv = getattr(g, 'current_user', None)
+    if rv == None:
+        email = get_user_email()
+        rv = MinespaceUser.query.unbound_unsafe().filter_by(email=email).filter_by(
```

Comment:
In Ruby on Rails, we'd do something like `@current_user ||= some_query.first()` to avoid running the query every time the method is called by storing the value in an instance variable. Is that being handled by the `if rv == None` check?

Reply:
Yes, g is the flask "current request" context. It's global to the current request, and dies when the current request is returned. So, this query will be called a maximum of once per API request, even if within the request you make several queries, and only if you are a proponent.